crates/algokit_transact/src/transactions/key_registration.rs

```diff
@@ -0,0 +1,110 @@
+//! which are used to register accounts online or offline for participation in Algorand consensus.
+
+use crate::transactions::common::TransactionHeader;
+use crate::utils::is_zero_opt;
+use derive_builder::Builder;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, skip_serializing_none, Bytes};
+
+/// Represents a key registration transaction that registers an account online or offline
+/// for participation in Algorand consensus.
+#[serde_as]
+#[skip_serializing_none]
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone, Builder)]
+#[builder(
+    name = "KeyRegistrationTransactionBuilder",
+    setter(strip_option),
+    build_fn(name = "build_fields")
+)]
+pub struct KeyRegistrationTransactionFields {
+    /// Common transaction header fields.
+    #[serde(flatten)]
+    pub header: TransactionHeader,
+
+    /// Root participation public key (32 bytes).
+    /// Required for online key registration.
+    #[serde(rename = "votekey")]
+    #[serde_as(as = "Option<Bytes>")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[builder(default)]
+    pub vote_key: Option<[u8; 32]>,
+
+    /// VRF public key (32 bytes).
+    /// Required for online key registration.
+    #[serde(rename = "selkey")]
+    #[serde_as(as = "Option<Bytes>")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[builder(default)]
+    pub selection_key: Option<[u8; 32]>,
+
+    /// State proof key (64 bytes).
+    /// Required for online key registration.
+    #[serde(rename = "sprfkey")]
+    #[serde_as(as = "Option<Bytes>")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[builder(default)]
+    pub state_proof_key: Option<[u8; 64]>,
+
+    /// First round for which the participation key is valid.
+    /// Required for online key registration.
+    #[serde(rename = "votefst")]
+    #[serde(skip_serializing_if = "is_zero_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub vote_first: Option<u64>,
+
+    /// Last round for which the participation key is valid.
+    /// Required for online key registration.
+    #[serde(rename = "votelst")]
+    #[serde(skip_serializing_if = "is_zero_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub vote_last: Option<u64>,
+
+    /// Key dilution for the 2-level participation key.
+    /// Required for online key registration.
+    #[serde(rename = "votekd")]
+    #[serde(skip_serializing_if = "is_zero_opt")]
+    #[serde(default)]
+    #[builder(default)]
+    pub vote_key_dilution: Option<u64>,
+
+    /// Mark account as non-reward earning.
+    /// Optional field that can be used with both online and offline registration.
+    #[serde(rename = "nonpart")]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    #[builder(default)]
+    pub non_participation: Option<bool>,
+}
+
+impl KeyRegistrationTransactionFields {
```

Comment:
I like this approach. Perhaps we have a `ValidTransaction` trait that has a `valid` function with a default implementation that does checks on the header and the calls a `valid_for_type` function that is called internally to do specific checks. 

As an example (untested, may have errors):

```rust
trait ValidTransaction {
    fn valid_for_type(&self) Result<(), String>;

    fn valid(&self) Result<(), String> {
        // Do checks on the header fields
        self.valid_for_type()
    }
}

impl ValidTransaction for KeyRegistrationTransaction {
    fn is_valid_for_type(&self) Result<(), String> {
        self.valid_offline()?
        self.valid_online()
    }
}
```

Reply:
@PatrickDinh Has done up an implementation of this in his PR https://github.com/algorandfoundation/algokit-core/pull/161/files
The naming is slightly different (`trait Validate` with `validate`) and there isn't any header validation as the header type mostly prevents invalid data being set, however that can be added like you have above very easily.

Reply:
@neilcampbell added trait Validate impl, ready for re review